### ISSUE TITLE
Regression: css fixed background cover image

### ISSUE THREAD

**Issue by markkimsal** [dompdf/dompdf#2775]
The solution for background image placement in #2422 stopped working in 1.2.0.
I'm not sure if this is a supported solution or not.

I have allow_url_fopen on on my system, and I'm including a local image off the local drive.  If there's some other option to set (like chroot) when upgrading from 1.1.1 to 1.2.0 I'm sorry that I missed it.

```html
<!DOCTYPE html>
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
  <style>
    @page {
        margin: 1in;
    }
    body {
      background-color: #00000066;
    }
    body:before {
      display: block;
      position: fixed;
      top: -1in; right: -1in; bottom: -1in; left: -1in;
      background-image: url(https://placekitten.com/500/500);
      background-size: cover;
      background-repeat: no-repeat;
      opacity: .2;
    }
  </style>
</head>
<body>
<p>
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam dictum risus eget faucibus porta. Quisque ut ipsum egestas tortor
 commodo pharetra. Sed elementum enim eu massa fringilla feugiat. Morbi tincidunt arcu malesuada accumsan porta. Nulla at
 eleifend enim. Phasellus fringilla malesuada orci at tincidunt. Nunc enim enim, cursus a tincidunt ultricies, luctus a enim. In posuere
 dictum erat, at tristique tortor lacinia at. Fusce viverra dolor ac augue condimentum tempus. Nullam gravida sit amet odio et viverra.
 Vestibulum ullamcorper at lectus pretium viverra.
</p>
</body>
</html>
````

**Comment by bsweeney** [comment]
Indeed, it looks like the optimization implemented in 415123e is to blame. I believe the new behavior is, mostly, correct? At least based on a quick look, but more thorough review is in order.

You can work around the issue by adding the "content" property to your generated element styling:

```
body:before {
  display: block;
  position: fixed;
  top: -1in; right: -1in; bottom: -1in; left: -1in;
  background-image: url(https://placekitten.com/500/500);
  background-size: cover;
  background-repeat: no-repeat;
  opacity: .2;
  content: "";
}
```

**Comment by markkimsal** [comment]
The solution in #2422 does have `content: "";` so I guess this could be categorized as `not a bug`

**Comment by bsweeney** [comment]
It does now, I edited the sample earlier so that anyone coming to it doesn't have a potentially broken work-around. Added a note to that post.

**Comment by bsweeney** [comment]
FYI there are some use cases that Dompdf appears to handle differently from web browsers so we'll review in context of the the specification to see if any adjustments to the optimization are necessary.

**Comment by Mellthas** [comment]
> Is this an optimization after decorating to clear out the child and sibling refs/pointers?

Yes, all rendered frames are disposed of after rendering each page:

https://github.com/dompdf/dompdf/blob/513a6cd872134394f3ffa2292570c038f431026d/src/FrameReflower/Page.php#L159-L172

The frame tree needs to be inspected during render, which can be done using `Dompdf::setCallbacks()`. Some of the existing tests utilize this method:

https://github.com/dompdf/dompdf/blob/513a6cd872134394f3ffa2292570c038f431026d/tests/DompdfTest.php#L116-L128

**Comment by Mellthas** [comment]
> Indeed, it looks like the optimization implemented in [415123e](https://github.com/dompdf/dompdf/commit/415123e5c5980508e80eab7321538fe2ecab2996) is to blame. I believe the new behavior is, mostly, correct? At least based on a quick look, but more thorough review is in order.

I think it is actually 56e45f9759e4fb1d0d0de0441ebbf55b22d5a3e4 to blame ;-). The new behavior is in line with the spec, as the default value for `content` is `normal`, which computes to `none` for before/after pseudo-elements; and `none` means that the pseudo-element should not be generated at all. See https://www.w3.org/TR/CSS21/generate.html#content.

**Comment by bsweeney** [comment]
It looks like this optimization has an unintended side effect in certain scenarios. For example, when the content property consists solely of a variable reference the pseudo-element is not rendered when the variable is not yet set. The most obvious impact is when using something like Font Awesome to display icons.
